### PR TITLE
BSTOR-85: Current order is kept in session storage

### PR DESCRIPTION
**Pull request by Elrhino**

### CODE REVIEW

File: client/src/main/java/com/arcbees/beestore/client/application/ApplicationPresenter.java [hunk 1]
```diff
@@ -108,6 +114,23 @@ protected void onBind() {
...
     }
 
+    @Override
+    protected void onReveal() {
+        populateCurrentOrderFromSession();
+        updateItemNumberTooltip();
+    }
+
+    private void populateCurrentOrderFromSession() {
+        List<ShoppingCartItem> items = sessionStorageHandler.getItems();
```
Comment (jclariviere) [inline client/src/main/java/com/arcbees/beestore/client/application/ApplicationPresenter.java]: You might want to do that in the Bootstrapper

File: client/src/main/java/com/arcbees/beestore/client/application/CurrentOrderImpl.java [hunk 2]
```diff
@@ -29,34 +30,70 @@
...
+        eventBus.addHandler(ShoppingCartQuantityChangeEvent.TYPE, this);
     }
 
     @Override
     public void addItem(final ShoppingCartItem item) {
-        ShoppingCartItem existingItemOfSameType = Iterables.tryFind(items, new Predicate<ShoppingCartItem>() {
+        ShoppingCartItem cartItem = getItemFromCart(item);
+        ShoppingCartItem sessionItem = getItemFromSession(item);
+
+        if (sessionItem != null && cartItem == null) {
```
Comment (jclariviere) [inline client/src/main/java/com/arcbees/beestore/client/application/CurrentOrderImpl.java]: This sounds clearer:

```
if (cartItem == null) {
    addItemToCart();
    if (sessionItem == null) {
        addItemToSession();
    }
}
```

File: client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    SessionStorageHandler(
+            SessionStorageKeyProvider keyProvider) {
+        this.keyProvider = keyProvider;
+    }
+
+    public void addToSessionStorage(ShoppingCartItem item) {
+        try {
+            sessionStorage.put(keyProvider.shoppingCartItemKey(item.getStorageKeyName()), item);
+        } catch (SerializationException e) {
+            e.printStackTrace();
```
Comment (jclariviere) [inline client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java]: Log it in the console

File: client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    public void updateFromSessionStorage(String keyName, int newQuantity) {
+        StorageKey<ShoppingCartItem> key = StorageKeyFactory.serializableKey(keyName);
+
+        try {
+            ShoppingCartItem itemToUpdate = sessionStorage.get(key);
+            itemToUpdate.setQuantity(newQuantity);
+            sessionStorage.put(key, itemToUpdate);
+        } catch (SerializationException e) {
+            e.printStackTrace();
```
Comment (jclariviere) [inline client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java]: Same here

File: client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        for (int i = 0; i < getSize(); i++) {
+            String keyValue = sessionStorage.key(i);
+            StorageKey<ShoppingCartItem> key = StorageKeyFactory.serializableKey(keyValue);
+
+            try {
+                ShoppingCartItem item = sessionStorage.get(key);
+                items.add(item);
+            } catch (SerializationException e) {
+                e.printStackTrace();
```
Comment (jclariviere) [inline client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java]: And here

File: client/src/main/java/com/arcbees/beestore/client/application/ShoppingCartItem.java [hunk 6]
```diff
@@ -57,4 +62,8 @@ public void setQuantity(int quantity) {
     public void addMore(int quantity) {
         this.quantity += quantity;
     }
+
+    public String getStorageKeyName() {
+        return productDto.getBrand().getValue() + "" + productDto.getProductType().name();
```
Comment (jclariviere) [inline client/src/main/java/com/arcbees/beestore/client/application/ShoppingCartItem.java]: `""`: I think there should be a space right? If not, that empty string does nothing.

File: client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.gwt.user.client.rpc.SerializationException;
+import com.google.inject.Inject;
+import com.seanchenxi.gwt.storage.client.StorageExt;
+import com.seanchenxi.gwt.storage.client.StorageKey;
+import com.seanchenxi.gwt.storage.client.StorageKeyFactory;
+
+public class SessionStorageHandler {
```
Comment (jclariviere) [inline client/src/main/java/com/arcbees/beestore/client/application/SessionStorageHandler.java]: You should use LocalStorage instead. SessionStorage is "per tab", it works when you reload, but not when you open a link in a new tab. See http://stackoverflow.com/a/5523174

File: common/src/main/java/com/arcbees/beestore/common/dto/ProductDto.java [hunk 8]
```diff
@@ -1,5 +1,9 @@
 package com.arcbees.beestore.common.dto;
```
Comment (olafleur) [inline common/src/main/java/com/arcbees/beestore/common/dto/ProductDto.java:1]: copyright notice

**Comment by olafleur** [comment]
Good job @jclariviere and @Elrhino ! LGTM.

File: common/src/main/java/com/arcbees/beestore/common/dto/ProductDto.java (hunk 8 above)
Comment (jclariviere) [inline common/src/main/java/com/arcbees/beestore/common/dto/ProductDto.java:1]: Will do it in another PR, there's multiple missing.
